external-crates/move/crates/move-compiler/src/sui_mode/linters/share_owned.rs

```diff
@@ -210,6 +219,108 @@ impl SimpleAbsInt for ShareOwnedVerifierAI {
     }
 }
 
+impl<'a> ShareOwnedVerifierAI<'a> {
+    fn can_hold_obj(&self, sp!(_, ty_): &Type) -> bool {
+        match ty_ {
+            Type_::Unit => false,
+            Type_::Single(st) => self.can_hold_obj_single(st),
+            Type_::Multiple(sts) => sts.iter().any(|st| self.can_hold_obj_single(st)),
+        }
+    }
+
+    fn can_hold_obj_single(&self, sp!(_, st_): &SingleType) -> bool {
+        match st_ {
+            SingleType_::Base(bt) | SingleType_::Ref(_, bt) => self.can_hold_obj_base(bt),
+        }
+    }
+
+    #[growing_stack]
+    fn can_hold_obj_base(&self, sp!(_, bt_): &BaseType) -> bool {
+        match bt_ {
+            // special case TxContext as not holding an object
+            BaseType_::Apply(_, sp!(_, tn), _)
+                if tn.is(SUI_ADDR_NAME, TX_CONTEXT_MODULE_NAME, TX_CONTEXT_TYPE_NAME) =>
+            {
+                false
+            }
+            // vector in value might have an object
+            BaseType_::Apply(
+                _,
+                sp!(_, TypeName_::Builtin(sp!(_, BuiltinTypeName_::Vector))),
+                bs,
+            ) => bs.iter().any(|b| self.can_hold_obj_base(b)),
+            // builtins cannot hold objects
+            BaseType_::Apply(_, sp!(_, TypeName_::Builtin(_)), _) => false,
+
+            BaseType_::Apply(_, sp!(_, TypeName_::ModuleType(m, n)), targs) => {
+                let m = *m;
+                let n = *n;
+                if self.sui_info().uid_holders.contains_key(&(m, n)) {
+                    return true;
+                }
+                let phantom_positions = phantom_positions(self.info, &m, &n);
+                phantom_positions
+                    .into_iter()
+                    .zip(targs)
+                    .filter(|(is_phantom, _)| !*is_phantom)
+                    .any(|(_, t)| self.can_hold_obj_base(t))
+            }
+            // any user defined type or type parameter is pessimistically assumed to hold an object
+            BaseType_::Param(_) => true,
+            BaseType_::Unreachable | BaseType_::UnresolvedError => false,
+        }
+    }
+
+    fn maybe_warn_share_owned(
+        &self,
+        context: &mut ExecutionContext,
+        loc: &Loc,
+        f: &ModuleCall,
+        args: Vec<Value>,
+    ) {
+        let Value::NotFreshObj(not_fresh_loc) = &args[0] else {
```

Comment:
Can `args[0]` here ever panic?

Reply:
Not at this point, but maybe above. I'll change